runtime/kubernetes/util.go

```diff
@@ -0,0 +1,206 @@
+package kubernetes
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/micro/go-micro/v3/logger"
+	"github.com/micro/go-micro/v3/runtime"
+	"github.com/micro/go-micro/v3/util/kubernetes/api"
+	"github.com/micro/go-micro/v3/util/kubernetes/client"
+)
+
+// getServices queries kubernetes for services. It gets information from both the pods and the
+// deployments
+func (k *kubernetes) getServices(opts ...client.GetOption) ([]*runtime.Service, error) {
+	// get the deployments
+	depList := new(client.DeploymentList)
+	d := &client.Resource{
+		Kind:  "deployment",
+		Value: depList,
+	}
+	if err := k.client.Get(d, opts...); err != nil {
+		return nil, err
+	}
+
+	srvMap := make(map[string]*runtime.Service, len(depList.Items))
+
+	// loop through the services and create a deployment for each
+	for _, kdep := range depList.Items {
+		srv := &runtime.Service{
+			Name:     kdep.Metadata.Labels["name"],
+			Version:  kdep.Metadata.Labels["version"],
+			Source:   kdep.Metadata.Labels["source"],
+			Metadata: kdep.Metadata.Annotations,
+		}
+
+		// this metadata was injected by the k8s runtime
+		delete(srv.Metadata, "name")
+		delete(srv.Metadata, "version")
+		delete(srv.Metadata, "source")
+
+		// parse out deployment status and inject into service metadata
+		if len(kdep.Status.Conditions) > 0 {
+			srv.Status = transformStatus(kdep.Status.Conditions[0].Type)
+			srv.Metadata["started"] = kdep.Status.Conditions[0].LastUpdateTime
+		} else {
+			srv.Status = runtime.Unknown
+		}
+
+		srvMap[resourceName(srv)] = srv
+	}
+
+	// get the pods from k8s
+	podList := new(client.PodList)
+	p := &client.Resource{
+		Kind:  "pod",
+		Value: podList,
+	}
+	if err := k.client.Get(p, opts...); err != nil {
+		logger.Errorf("Error fetching pods: %v", err)
+		return nil, nil
+	}
+
+	for _, item := range podList.Items {
```

Comment:
minor: if/when we support multiple instances for a service this will only take the status from one pod

Reply:
Good point, I think this will require a wider rethink within the package.

Reply:
func transformStatus(depStatus string)  has so many case, if we can use the map optimization.